src/lib/crypto/kimchi_backend/gadgets/xor.ml

```diff
@@ -0,0 +1,333 @@
+        let input2_bits = Common.pad_upto ~length ~value:false input2_bits in
+
+        (* Pad with more zeros until the length is a multiple of 4*n for n-bit length lookup table *)
+        let pad_length =
+          if length mod (4 * len_xor) <> 0 then
+            length + (4 * len_xor) - (length mod (4 * len_xor))
+          else length
+        in
+        let input1_bits =
+          Common.pad_upto ~length:pad_length ~value:false input1_bits
+        in
+        let input2_bits =
+          Common.pad_upto ~length:pad_length ~value:false input2_bits
+        in
+
+        (* Xor list of bits to obtain output of the xor *)
+        let output_bits =
+          List.map2_exn input1_bits input2_bits ~f:(fun b1 b2 ->
+              Bool.(not (equal b1 b2)) )
+        in
+        (* Recursively build Xor gadget *)
+        bxor_rec input1_bits input2_bits output_bits pad_length len_xor ;
+
+        (* Convert back to field *)
+        Field.Constant.project output_bits )
+  in
+
+  output_xor
+
+(* Xor of 16 bits *)
+let bxor16 (type f)
+    (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+    (input1 : Circuit.Field.t) (input2 : Circuit.Field.t) : Circuit.Field.t =
+  bxor (module Circuit) input1 input2 16 4
+
+(* Xor of 64 bits *)
+let bxor64 (type f)
+    (module Circuit : Snarky_backendless.Snark_intf.Run with type field = f)
+    (input1 : Circuit.Field.t) (input2 : Circuit.Field.t) : Circuit.Field.t =
+  bxor (module Circuit) input1 input2 64 4
+
+let%test_unit "xor gadget" =
+  (* Import the gadget test runner *)
+  let open Kimchi_gadgets_test_runner in
+  (* Initialize the SRS cache. *)
+  let () =
+    try Kimchi_pasta.Vesta_based_plonk.Keypair.set_urs_info [] with _ -> ()
+  in
+
+  (* Helper to test Xor gadget
+     *   Inputs operands and expected output: left_input xor right_input
+     *   Returns true if constraints are satisfied, false otherwise.
+  *)
+  let test_xor left_input right_input output_xor length : unit =
+    let _proof_keypair, _proof =
+      Runner.generate_and_verify_proof (fun () ->
+          let open Runner.Impl in
+          (* Set up snarky variables for inputs and output *)
+          let left_input =
+            exists Field.typ ~compute:(fun () ->
+                Field.Constant.of_string left_input )
+          in
+          let right_input =
+            exists Field.typ ~compute:(fun () ->
+                Field.Constant.of_string right_input )
+          in
+          let output_xor =
+            exists Field.typ ~compute:(fun () ->
+                Field.Constant.of_string output_xor )
+          in
+          (* Use the xor gate gadget *)
+          let result =
+            bxor (module Runner.Impl) left_input right_input length 4
+          in
+          Field.Assert.equal output_xor result ;
+          (* Pad with a "dummy" constraint b/c Kimchi requires at least 2 *)
+          Boolean.Assert.is_true (Field.equal result result) )
```

Comment:
I guess you may not need this padding, since your xor is always more than one gate?

Reply:
oh okay, yes it is always more than one row